[CBRD-24834] Utility also works with invalid cubrid_hosts.conf file configuration

eido5 opened the pull request:
http://jira.cubrid.org/browse/CBRD-24834

**Description**
* in addition to #3740,
* fix to do same style action of cub_master with glibc host lookup when cubrid_hosts.conf looks like:
```
127.0.0.1 localhost
192.168.240
```

**Remarks**

eido5 committed:
[CBRD-24834] Utility also works with invalid cubrid_hosts.conf file configuration

eido5 committed:
Eliminate failure case when the duplicated ip address is written in the cubrid_hosts.conf

eido5 commented:
@kisoo-han 
`(hostent_Cache[cache_idx] = hostent_alloc (ipaddr, hostname)) == NULL`
This code and inside **hostent_alloc ()** occurs NULL return value when the malloc() is failed.
So, this is not necessary to be fixed.

Is it right what your comment is?

eido5 committed:
Exchange code to execute print like glibc

eido5 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-24834

kisoo-han commented on src/connection/host_lookup.c:
```diff
@@ -346,65 +345,35 @@ load_hosts_file ()
 
-load_fail_phase:
+load_end_phase:
+  if (fp != NULL)
+    fclose (fp);
```
```suggestion
  if (fp != NULL)
    {
      fclose (fp);
    }      
```

eido5 replied:
I will fix it. Thank you.

eido5 committed:
Add brace between fclose()

eido5 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-24834

kisoo-han commented on src/connection/host_lookup.c:
```diff
@@ -303,37 +304,35 @@ load_hosts_file ()
+		  continue;
 		}
-	      else if (ip_format_check (temp_token) == true)
+	      else if (hostname_format_check (temp_token, str_len) == false)
```
what about change function name as (is_valid_hostname): _trivial_
else if (**is_valid_hostname** (temp_token, str_len) == false)

eido5 replied:
If then, I will change "ip_format_check" to "is_valid_ip" either.

kisoo-han commented on src/connection/host_lookup.c:
```diff
@@ -462,6 +433,46 @@ ip_format_check (char *ip_addr)
+
+  int char_num = 0;
+
+  if (!((hostname[0] >= 'A' && hostname[0] <= 'Z') ^ (hostname[0] >= 'a' && hostname[0] <= 'z')))
```
For the readability, I propose:

`if (!isalpha (hostname[0]))`

eido5 replied:
It will be better. I will apply it.

kisoo-han commented on src/connection/host_lookup.c:
```diff
@@ -462,6 +433,46 @@ ip_format_check (char *ip_addr)
+    }
+
+
+  if (!((hostname[str_len - 1] >= 'A' && hostname[str_len - 1] <= 'Z') ^
+	(hostname[str_len - 1] >= 'a' && hostname[str_len - 1] <= 'z') ^
+	(hostname[str_len - 1] >= '0' && hostname[str_len - 1] <= '9')))
```
For the same reason,
if (!isalnum (hostname[str_len - 1]))

kisoo-han commented on src/connection/host_lookup.c:
```diff
@@ -462,6 +433,46 @@ ip_format_check (char *ip_addr)
+
+  for (char_num = 1; char_num < str_len - 1; char_num++)
+    {
+      if (!((hostname[char_num] >= 'A' && hostname[char_num] <= 'Z') ^
+	    (hostname[char_num] >= 'a' && hostname[char_num] <= 'z') ^
+	    (hostname[char_num] >= '0' && hostname[char_num] <= '9') ^
+	    (hostname[char_num] == '-') ^ (hostname[char_num] == '.')))
```
if (!(isalnum (hostname[char_num]) || hostname[char_num] == '-' || hostname[char_num] == '.'))

eido5 committed:
Change function name and improve readability

eido5 committed:
Fix additional coniditions

eido5 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-24834

kisoo-han commented:
It is recommended to divide the message for hostname lookup error into two types according to the value of the **use_user_hosts** system parameter as follows. Reference of "/etc/hosts" should be modified in case of value of use_user_hosts is true.
* Cannot find ("(unknown)") in /etc/hosts. Please check permissions of the file or if there is the hostname.
* Cannot find ("(unknown)") in **$CUBRID/conf/cubrid_hosts.conf**. Please check permissions of the file or if there is the hostname.

eido5 committed:
Edit error message in the tcp.c

eido5 commented:
I edited error message format only in the "msg/en_US/cubrid.msg".
If it is edited well, please reply on this comment. and then I will edit other languages one.
@kisoo-han

kisoo-han commented:
Good. Go ahead!

iOS용 Outlook<https://aka.ms/o0ukef> 다운로드
________________________________
보낸 사람: eido5 ***@***.***>
보낸 날짜: Thursday, June 15, 2023 8:46:00 PM
받는 사람: CUBRID/cubrid ***@***.***>
참조: 한기수 ***@***.***>; Mention ***@***.***>
제목: Re: [CUBRID/cubrid] [CBRD-24834] Utility also works with invalid cubrid_hosts.conf file configuration (PR #4414)


I edited error message format only in the "msg/en_US/cubrid.msg".
If it is edited well, please reply on this comment. and then I will edit other languages one.
@kisoo-han<https://github.com/kisoo-han>

—
Reply to this email directly, view it on GitHub<https://github.com/CUBRID/cubrid/pull/4414#issuecomment-1592888796>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AEWKVVQLUSRZCPWL4XMMXMTXLLYXRANCNFSM6AAAAAAY7D42GU>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

eido5 committed:
Edit error message and delete deprecated message

kisoo-han commented on src/connection/host_lookup.c:
```diff
@@ -346,74 +345,46 @@ load_hosts_file ()
 	  /*not duplicated hostname, IP address or not duplicated hostname and duplicated IP address */
-	  if ((user_host_Map.find (hostname) == user_host_Map.end ()))
+	  if ((user_host_Map.find (hostname) == user_host_Map.end ())
+	      && (user_host_Map.find (ipaddr) == user_host_Map.end ()))
```
This function can replace alias. I'd like to keep the original.

```
/etc/hosts
192.168.2.34 dev1 dev-1
```

```
cubrid_hosts.conf
192.168.2.34 dev1
192.168.2.34 dev-1
```

eido5 replied:
I knew that etc/hosts saves first of one if there are duplicated ip address.
But in this case, this hash map function saves latest one. So, I wrote above code.

So, when you find '192.168.2.34' and dev-1 is returned.

kisoo-han replied:
Ok, let's keep change.

eido5 commented on src/connection/host_lookup.c:
```diff
@@ -298,42 +299,40 @@ load_hosts_file ()
+	      str_len = strlen (temp_token);
+	      if (str_len > HOSTNAME_LEN - 1)
 		{
-		  er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_UHOST_HOST_NAME_TOO_LONG, 3, token, line_num,
```
Should I roll back this coded?
Because user cannot know which one is the problem.
@kisoo-han

kisoo-han replied:
Ok, let's keep change.

eido5 replied:
Ok. I will recover it.

kisoo-han replied:
Considering compatibility with glibc, we don't need to give these error messages at library level. If necessary, it is better to do it in the function that call host lookup.
So, do not roll back.

eido5 replied:
It does not print error into the stdout. Just record in the log. 
If you mean that log recording is also banned. I will keep it.

eido5 commented:
I think we should finish task about what we are requested or we fixed about the error message in this PR.
(at least **direct related with** host look up)

1. fprintf (stderr, "unknown host : %s\n", hostname); This message is refered in the last meeting. So, we need fix it because we fixed the "er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_GAI_ERROR, 1, hostname, HOSTS_FILE);" 
2. Error log recording (**not printed to the stdout**) should be fixed about the over 256 length hostname.

I think above two things what we did similar in this PR, So we need to finish these.
If not, It remains the code integrity violation.
And we can decide beyond master or broker error message can be handled this fig release.

eido5 committed:
Set additional error argument; etc/hosts or cubrid_hosts.conf

eido5 committed:
Merge remote-tracking branch 'upstream/develop' into CBRD-24834

kisoo-han reviewed (approved):
please merge and mark CBRD-24834 as resolved.
